src/mesh/amr_loadbalance.cpp

```diff
@@ -70,73 +72,94 @@ void Mesh::LoadBalancingAndAdaptiveMeshRefinement(ParameterInput *pin) {
   return;
 }
 
-//----------------------------------------------------------------------------------------
-// \!fn void Mesh::CalculateLoadBalance(double *clist, int *rlist, int *slist,
-//                                      int *nlist, int nb)
-// \brief Calculate distribution of MeshBlocks based on the cost list
-
-void Mesh::CalculateLoadBalance(double *clist, int *rlist, int *slist, int *nlist,
-                                int nb) {
-  std::stringstream msg;
-  double real_max = std::numeric_limits<double>::max();
-  double totalcost = 0, maxcost = 0.0, mincost = (real_max);
-
-  for (int i = 0; i < nb; i++) {
-    totalcost += clist[i];
-    mincost = std::min(mincost, clist[i]);
-    maxcost = std::max(maxcost, clist[i]);
-  }
-
-  int j = (Globals::nranks)-1;
-  double targetcost = totalcost / Globals::nranks;
-  double mycost = 0.0;
+// Private routines
+namespace {
+/**
+ * @brief This routine assigns blocks to ranks by attempting to place index-contiguous
+ * blocks of equal total cost on each rank.
+ *
+ * @param costlist (Input) A map of global block ID to a relative weight.
+ * @param ranklist (Output) A map of global block ID to ranks.
+ */
+void AssignBlocks(std::vector<double> const &costlist, std::vector<int> &ranklist) {
```

Comment:
ranklist should be returned by AssignBlocks and not be brought in as a reference. The only place AssignBlocks is called is within the CalculateLoadBalance method. It should also probably be returned by that method instead of passed in as a reference as well. Every time rank list is passed into CalculateLoadBalance it has nothing stored in it. 
Optional:

This might also be a different issue, I would have CalcualteLoadBalance return some struct that has the results that are initialized, I do not like passing in objects as references if you don't actually need their contents. In this case rank list is always empty when it is passed in to CalculateLoadBalance.

Reply:
Regarding `AssignBlocks`, I disagree. The reason to pass it in by reference is because it's a `std::vector`. It's best practice not to return vectors by value, since they're essentially arrays. And they could be substantial in size, `nblocks`, which could be O(100) or more.

Same with with `CalculateLoadBalance`, although I agree that the the arguments could be consolidated into a `struct`, I think that struct should still be passed by reference. And since it's only 3 or 4 args, I'm not sure it's worth it.